src/lua_generator.cpp

```diff
@@ -0,0 +1,769 @@
+        requires_(),
+        current_block_(),
+        flatc_version_(flatc_version) {
+    static const char *const keywords[] = {
+      "and",      "break",  "do",   "else", "elseif", "end",  "false", "for",
+      "function", "goto",   "if",   "in",   "local",  "nil",  "not",   "or",
+      "repeat",   "return", "then", "true", "until",  "while"
+    };
+    keywords_.insert(std::begin(keywords), std::end(keywords));
+  }
+
+  GeneratorStatus generate(const r::Schema *schema) override {
+    if (!generate_enums(schema->enums())) { return GeneratorStatus::FAILED; }
+    if (!generate_objects(schema->objects(), schema->root_table())) {
+      return GeneratorStatus::FAILED;
+    }
+    return GeneratorStatus::OK;
+  }
+
+ protected:
+  bool generate_enums(
+      const flatbuffers::Vector<flatbuffers::Offset<r::Enum>> *enums) {
+    for (auto it = enums->cbegin(); it != enums->cend(); ++it) {
+      auto enum_def = *it;
+      start_code_block(enum_def);
+      std::string ns;
+      const std::string enum_name = generate_enum(enum_def, ns);
+      emit_code_block(enum_name, ns, enum_def->declaration_file()->str());
+    }
+    return true;
+  }
+
+  std::string generate_enum(const r::Enum *enum_def, std::string &ns) {
+    const std::string enum_name =
+        normalize_name(denamespace(enum_def->name(), ns));
+
+    append_comments(enum_def->documentation());
+    append_line("local " + enum_name + " = {");
+    {
+      // TODO(derekbailey): It would be cool if this would auto dedent on
+      // leaving scope.
+      indent();
+      for (auto it = enum_def->values()->cbegin();
+           it != enum_def->values()->cend(); ++it) {
+        const auto enum_val = *it;
+        append_comments(enum_val->documentation());
+        append_line(normalize_name(enum_val->name()) + " = " +
+                    std::to_string(enum_val->value()) + ",");
+      }
+      dedent();
+    }
+    append_line("}");
+    append_line();
+    return enum_name;
+  }
+
+  bool generate_objects(
+      const flatbuffers::Vector<flatbuffers::Offset<r::Object>> *objects,
+      const r::Object *root_object) {
+    for (auto it = objects->cbegin(); it != objects->cend(); ++it) {
+      auto object_def = *it;
+      start_code_block(object_def);
+      std::string ns;
+      // Register the main flatbuffers module.
+      register_requires("flatbuffers", "flatbuffers");
+      const std::string object_name =
+          generate_object(object_def, ns, object_def == root_object);
+      emit_code_block(object_name, ns, object_def->declaration_file()->str());
+    }
+    return true;
+  }
+
+  std::string generate_object(const r::Object *object_def, std::string &ns,
+                              bool is_root_object) {
+    const std::string object_name =
+        normalize_name(denamespace(object_def->name(), ns));
+
+    append_comments(object_def->documentation());
+    append_line("local " + object_name + " = {}");
+    append_line("local mt = {}");
```

Comment:
This is going to be very much a preference thing, but this feels like a very verbose style of codegen to me. That would be ok if it was just for Lua, but by putting these functions in the base class you're implying that all such generators should use this style.

What should we use instead? Frankly, if its just down to me, I would not have such functions at all, and just `+=` raw strings, including embedded indentation spaces. I think all this procedural indenting is overabstraction for any language that has a clear code style. Even for C++ we just output Google Style code, so it doesn't matter. And not having all these function calls in between makes it all much easier to read.

If that is too basic for you, maybe something like what the C++ generator does, with overloaded `+=` etc.. but not a fan of that either.

Reply:
Noted. I'll moved out of base class.

Reply:
I do prefer the more verbose style, especially if performance is not an issue. Which it isn't during code gen. The compiler is just going to inline all the method calls anyways.

Reply:
I'll leave it for now and we can refactor after this is in.

Reply:
While we're bike shedding, my 2c is that I like the += thing. I also try to use methods like `ForAllFieldsInTable` or `ForAllVariantsInEnum` which take a lambda and handles indentation. This is neat since the code in the lambda will be indented anyway. If the ForAllX method did not handle indentation, then the raw string would seem double indented. This is of course pretty far on the DRY/overengineered side of the spectrum

Reply:
Are you suggesting placing `ForAllFieldsInTable` and `ForAllVariantsInEnum` in the base class and then the implementing classes just provide a custom lambda?

Reply:
I went ahead and refactored it all to use the `code +=` style of code gen.

I took @CasperN idea to define those functions that took a lambda, as that hid away the details of how we were looping over the vectors.